### PR TITLE
Support adding users to T-SQL db_owner database role

### PR DESCRIPTION
**Pull request by thephantomthief**
### Description

This commit adds support to alter the db_owner role and add/drop users from it.

Whenever a new user, say u1, is added to db_owner role, we will internally create a new role u1_obj which will be owner of all the objects of u1. By doing so, we are able to work around the cyclic dependency of role membership in postgres. If u1 creates new objects, using the pre-existing post object creation hook, we will reassign ownership of u1's objects to u1_obj.

Similarly, if u1 is dropped from db_owner role, we will drop the u1_obj role and reassign ownership u1's objects from u1_obj role back to u1.

Task: BABEL-4899
Signed-off-by: Sharu Goel <goelshar@amazon.com>

### Test Scenarios Covered ###
* **Use case based -** Yes

* **Boundary conditions -** N/A

* **Arbitrary inputs -** N/A

* **Negative test cases -** Yes

* **Minor version upgrade tests -** Yes

* **Major version upgrade tests -** Yes

* **Performance tests -** N/A

* **Tooling impact -** N/A

* **Client tests -** N/A

### Check List
- [x] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is under the terms of the Apache 2.0 and PostgreSQL licenses, and grant any person obtaining a copy of the contribution permission to relicense all or a portion of my contribution to the PostgreSQL License solely to contribute all or a portion of my contribution to the PostgreSQL open source project.

For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/babelfish-for-postgresql/babelfish_extensions/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Comment by tanscorpio7** [comment]
Who will own schema if user runs
```
CREATE SCHEMA s1 AUTHORIZATION l1
```
and l1 is a member of db_owner. Will the internal role own the schema ?

**Comment by tanscorpio7** [comment]
The current grant on schema logic assumes only dbo and schema_owner can create objects in the schema.
Now any role which is a member of db_owner can create objects in any schema.

Question is will objects created by members of db_owner adhere to grants on schema ?
```
grant select on schema::s1 to l2
```
then as a member of db_owner create in that schema
```
create table s1.t (id int) <-- can l2 select on this ?
```

File: contrib/babelfishpg_tsql/src/rolecmds.c [hunk 1]
```diff
@@ -2605,3 +2664,446 @@ bbf_is_member_of_role_nosuper(PG_FUNCTION_ARGS)
...
+	char		*rolname = NULL;
+	char 		*obj_rolname = NULL;
+	char		*cur_db_name = NULL;
+	Oid		role_oid = GetUserId();
+	List		*parsetree_list;
+	Node		*n;
+	PlannedStmt 	*wrapper;
+
+	/* TSQL specific behavior */
+	if (sql_dialect != SQL_DIALECT_TSQL)
```
Comment (tanscorpio7) [inline contrib/babelfishpg_tsql/src/rolecmds.c]: Should we also check if current connection is TDS CONN ?

File: contrib/babelfishpg_tsql/src/rolecmds.c [hunk 2]
```diff
@@ -2605,3 +2664,446 @@ bbf_is_member_of_role_nosuper(PG_FUNCTION_ARGS)
...
+		return;
+
```
Comment (tanscorpio7) [inline contrib/babelfishpg_tsql/src/rolecmds.c:3095]: We probably should have `CommandCounterIncrement();` here

File: contrib/babelfishpg_tsql/src/hooks.c [hunk 3]
```diff
@@ -2965,6 +2965,9 @@ bbf_object_access_hook(ObjectAccessType access, Oid classId, Oid objectId, int s
 
 	if (access == OAT_POST_CREATE && classId == ProcedureRelationId)
 		revoke_func_permission_from_public(objectId);
+
+	if (access == OAT_POST_CREATE)
+		change_object_owner_if_db_owner();
```
Comment (tanscorpio7) [inline contrib/babelfishpg_tsql/src/hooks.c:2983]: Since we are doing this way, we could add test cases for all the pg_catalog which contains some metadata related to ownerId and the newly created object.

For example when during create schema we record dependency on owner and default acls for the schema which also includes the ownerId, so we can add tests for both.

Not expecting to see any issues here, but would be good to have test cases. Thanks

File: contrib/babelfishpg_tsql/sql/ownership.sql [hunk 4]
```diff
@@ -440,6 +440,10 @@ SELECT pg_catalog.pg_extension_config_dump('sys.babelfish_authid_login_ext', '')
 SELECT pg_catalog.pg_extension_config_dump('sys.babelfish_authid_user_ext', '');
 SELECT pg_catalog.pg_extension_config_dump('sys.babelfish_schema_permissions', '');
 
+CREATE OR REPLACE FUNCTION sys.bbf_is_role_member(member NAME, rolename NAME)
+RETURNS BOOLEAN
+AS 'babelfishpg_tsql', 'bbf_is_role_member' LANGUAGE C;
+
```
Comment (tanscorpio7) [inline contrib/babelfishpg_tsql/sql/ownership.sql:446]: Is there any possibility of merging this with `bbf_is_member_of_role_nosuper` ?

**Review by tanscorpio7 — CHANGES_REQUESTED** [review]

File: contrib/babelfishpg_tsql/src/dbcmds.c [hunk 5]
```diff
@@ -818,19 +820,27 @@ drop_bbf_db(const char *dbname, bool missing_ok, bool force_drop)
 				 errmsg("Cannot drop database \"%s\" because it is currently in use", dbname)));
 
 	/* Set current user to session user for dropping permissions */
+	prev_current_user_id = GetSessionUserId();
```
Comment (HarshLunagariya) [inline contrib/babelfishpg_tsql/src/dbcmds.c]: Better naming would be prev_session_user_id

File: contrib/babelfishpg_tsql/src/pl_handler.c [hunk 6]
```diff
@@ -3069,7 +3069,15 @@ bbf_ProcessUtility(PlannedStmt *pstmt,
...
-							create_bbf_authid_user_ext(stmt, isuser, isuser, from_windows);
+
+							/*
+							 * If the role is created internally as part of ALTER ROLE
+							 * db_owner ADD MEMBER ... statement, we should not add this to
+							 * our babelfish catalog. These roles are meant to be internal
+							 * and not be visible to customer from Babelfish endpoint.
+							 */
+							if (strcmp(queryString, "(ALTER ROLE ADD )") != 0)
+								create_bbf_authid_user_ext(stmt, isuser, isuser, from_windows);
```
Comment (HarshLunagariya) [inline contrib/babelfishpg_tsql/src/pl_handler.c]: We are only creating _obj role when adding any user as member of db_owner , right? Then should we have some kind of metadata just to rely on it to decide -  for which user, have we already created _obj_role and for which user, we haven't?

File: contrib/babelfishpg_tsql/src/pl_handler.c [hunk 7]
```diff
@@ -3955,13 +3963,19 @@ bbf_ProcessUtility(PlannedStmt *pstmt,
...
 						else
-							standard_ProcessUtility(pstmt, queryString, readOnlyTree, context, params,
+						{
+							if (prev_ProcessUtility)
+								prev_ProcessUtility(pstmt, queryString, readOnlyTree, context, params,
 													queryEnv, dest, qc);
-
+							else
+								standard_ProcessUtility(pstmt, queryString, readOnlyTree, context, params,
+														queryEnv, dest, qc);
```
Comment (HarshLunagariya) [inline contrib/babelfishpg_tsql/src/pl_handler.c]: nit: Let's utilise - `call_prev_ProcessUtility()` here

File: contrib/babelfishpg_tsql/src/rolecmds.c [hunk 8]
```diff
@@ -1603,48 +1606,93 @@ alter_bbf_authid_user_ext(AlterRoleStmt *stmt)
...
+			ereport(ERROR,
+					(errcode(ERRCODE_SYNTAX_ERROR),
+					 errmsg("Expected 1 statement but got %d statements after parsing",
+							list_length(parsetree_list))));
+		else if (is_db_owner_member && list_length(parsetree_list) != 2)
 			ereport(ERROR,
 					(errcode(ERRCODE_SYNTAX_ERROR),
-					 errmsg("Expected 1 statement but get %d statements after parsing",
+					 errmsg("Expected 2 statement but got %d statements after parsing",
 							list_length(parsetree_list))));
```
Comment (HarshLunagariya) [inline contrib/babelfishpg_tsql/src/rolecmds.c]: Let's refactor this code to avoid duplication. Instead of 1/2 let's maintain a variable.

File: contrib/babelfishpg_tsql/src/rolecmds.c [hunk 9]
```diff
@@ -1838,13 +1887,21 @@ check_alter_role_stmt(GrantRoleStmt *stmt)
...
+	    !has_privs_of_role(GetUserId(), db_owner))
 		ereport(ERROR,
 				(errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
 				 errmsg("Cannot alter the role '%s', because it does not exist or you do not have permission.", original_user_name)));
 
+	/*
+	 * Members of db_owner role can ALTER ANY ROLE
+	 */
+	if (has_privs_of_role(GetSessionUserId(), db_owner))
+		return;
```
Comment (HarshLunagariya) [inline contrib/babelfishpg_tsql/src/rolecmds.c]: Shouldn't it be CurrentUser?

File: contrib/babelfishpg_tsql/src/rolecmds.c [hunk 10]
```diff
@@ -2605,3 +2664,446 @@ bbf_is_member_of_role_nosuper(PG_FUNCTION_ARGS)
...
+char*
+get_obj_role(const char *rolname)
+{
+	StringInfoData rolname_obj;
+
+	initStringInfo(&rolname_obj);
+
+	appendStringInfoString(&rolname_obj, rolname);
+	appendStringInfoString(&rolname_obj, "_obj");
+
```
Comment (HarshLunagariya) [inline contrib/babelfishpg_tsql/src/rolecmds.c:2812]: Let's make sure we add testcases for long username and database name.

File: contrib/babelfishpg_tsql/src/rolecmds.c [hunk 11]
```diff
@@ -2605,3 +2664,446 @@ bbf_is_member_of_role_nosuper(PG_FUNCTION_ARGS)
...
+	if (is_member_of_role(roleid, db_owner_id) && (roleid != InvalidOid))
+	{
+		StringInfoData	query;
+		List		*parsetree_list;
+		Node		*n;
+		PlannedStmt	*wrapper;
+
+		initStringInfo(&query);
+		appendStringInfo(&query, "DROP ROLE dummy; ");
```
Comment (HarshLunagariya) [inline contrib/babelfishpg_tsql/src/rolecmds.c:3215]: Should we execute `DROP OWNED BY` statement here as well?

File: contrib/babelfishpg_tsql/src/rolecmds.c [hunk 12]
```diff
@@ -2605,3 +2664,446 @@ bbf_is_member_of_role_nosuper(PG_FUNCTION_ARGS)
...
+	if (obj_rolname)
+		pfree(obj_rolname);
+}
+
+PG_FUNCTION_INFO_V1(bbf_is_role_member);
+
+/*
+ * The bbf_is_role_memeber function will check if a user (u1) is member of a role (r1):
+ * - If role is not a member of db_owner, we will check if u1 is member of r1 role
+ * - If role is a member of db_owner, we will check if u1 is member of r1_obj role
```
Comment (HarshLunagariya) [inline contrib/babelfishpg_tsql/src/rolecmds.c:3022]: > If role is a member of db_owner, we will check if u1 is member of r1_obj role
We need to test this thoroughly in terms of metadata visibility as well.

File: contrib/babelfishpg_tsql/src/rolecmds.c (hunk 1 above)
Comment (HarshLunagariya) [reply contrib/babelfishpg_tsql/src/rolecmds.c]: +1

File: contrib/babelfishpg_tsql/src/rolecmds.c [hunk 13]
```diff
@@ -2605,3 +2664,446 @@ bbf_is_member_of_role_nosuper(PG_FUNCTION_ARGS)
...
+
+	rolname = GetUserNameFromId(role_oid, true);
+
+	if (!rolname)
+		return;
+
+	if (!user_exists_for_db(cur_db_name, rolname))
+		return;
+
+	if (!is_member_of_role(role_oid, get_role_oid(get_db_owner_name(cur_db_name), true)))
```
Comment (HarshLunagariya) [inline contrib/babelfishpg_tsql/src/rolecmds.c]: missing ok is true here, is it expected?

File: contrib/babelfishpg_tsql/src/rolecmds.c [hunk 14]
```diff
@@ -2605,3 +2664,446 @@ bbf_is_member_of_role_nosuper(PG_FUNCTION_ARGS)
...
+
+		schemas++;
+	}
+
+	systable_endscan(sscan);
+	table_close(rel, AccessShareLock);
+
+	/* Ownership of rest of the objects will be taken care of by REASSIGN OWNED */
+	appendStringInfoString(&query, "REASSIGN OWNED BY dummy TO dummy; ");
+	appendStringInfoString(&query, "DROP ROLE dummy; ");
```
Comment (HarshLunagariya) [inline contrib/babelfishpg_tsql/src/rolecmds.c]: REASSIGN OWNED BY doesn't cover schema object?

File: contrib/babelfishpg_tsql/src/rolecmds.c [hunk 15]
```diff
@@ -2605,3 +2664,446 @@ bbf_is_member_of_role_nosuper(PG_FUNCTION_ARGS)
...
+	ScanKeyData	skey;
+	Relation	rel;
+
+	initStringInfo(&query);
+
+	appendStringInfoString(&query, "CREATE ROLE dummy; ");
+	appendStringInfoString(&query, "GRANT dummy TO dummy; ");
+
+	/* We will first transfer ownership of all the schemas owned by user */
+	rel = table_open(NamespaceRelationId, AccessShareLock);
```
Comment (HarshLunagariya) [inline contrib/babelfishpg_tsql/src/rolecmds.c:2845]: RowExclusiveLock?

File: contrib/babelfishpg_tsql/src/rolecmds.c [hunk 16]
```diff
@@ -2605,3 +2664,446 @@ bbf_is_member_of_role_nosuper(PG_FUNCTION_ARGS)
...
+
+	initStringInfo(&query);
+
+	appendStringInfoString(&query, "REVOKE dummy FROM dummy; ");
+	appendStringInfoString(&query, "REVOKE dummy FROM dummy; ");
+	appendStringInfoString(&query, "GRANT dummy TO dummy; ");
+	appendStringInfoString(&query, "REVOKE dummy FROM dummy; ");
+
+	/* We will first transfer ownership of all the schemas owned by user */
+	rel = table_open(NamespaceRelationId, AccessShareLock);
```
Comment (HarshLunagariya) [inline contrib/babelfishpg_tsql/src/rolecmds.c:2945]: RowExclusiveLock?

File: contrib/babelfishpg_tsql/src/rolecmds.c (hunk 1 above)
Comment (thephantomthief) [reply contrib/babelfishpg_tsql/src/rolecmds.c]: Added check for TDS connection

File: contrib/babelfishpg_tsql/src/rolecmds.c (hunk 2 above)
Comment (thephantomthief) [reply contrib/babelfishpg_tsql/src/rolecmds.c:3095]: Done

File: contrib/babelfishpg_tsql/src/hooks.c [hunk 17]
```diff
@@ -5719,3 +5725,63 @@ handle_grantstmt_for_dbsecadmin(ObjectType objType, Oid objId, Oid ownerId,
...
+			nsp_owner = nsptup->nspowner;
+
+		if (ownerId != nsp_owner)
+		{
+			Oid 	db_owner = get_db_owner_oid(db_name, false);
+			Oid 	schema_db_id = get_dbid_from_physical_schema_name(NameStr(nsptup->nspname), false);
+
+			if (schema_db_id == get_cur_db_id() &&
+				has_privs_of_role(GetUserId(), db_owner))
+				ownerId = nsp_owner;
```
Comment (tanscorpio7) [inline contrib/babelfishpg_tsql/src/hooks.c]: Let's skip this if current_user is dbo. For dbo we can fix it with BABEL-5361

**Review by tanscorpio7 — CHANGES_REQUESTED** [review]

File: contrib/babelfishpg_tsql/src/hooks.c (hunk 17 above)
Comment (thephantomthief) [reply contrib/babelfishpg_tsql/src/hooks.c]: Done

File: contrib/babelfishpg_tsql/src/dbcmds.c (hunk 5 above)
Comment (thephantomthief) [reply contrib/babelfishpg_tsql/src/dbcmds.c]: Done

File: contrib/babelfishpg_tsql/src/pl_handler.c (hunk 6 above)
Comment (thephantomthief) [reply contrib/babelfishpg_tsql/src/pl_handler.c]: As discussed offline, if member of db_owner is dropped from db_owner role or the member user is dropped from database, we will also drop the corresponding _obj role

File: contrib/babelfishpg_tsql/src/pl_handler.c (hunk 7 above)
Comment (thephantomthief) [reply contrib/babelfishpg_tsql/src/pl_handler.c]: Done

File: contrib/babelfishpg_tsql/src/rolecmds.c (hunk 8 above)
Comment (thephantomthief) [reply contrib/babelfishpg_tsql/src/rolecmds.c]: Done

File: contrib/babelfishpg_tsql/src/rolecmds.c (hunk 9 above)
Comment (thephantomthief) [reply contrib/babelfishpg_tsql/src/rolecmds.c]: Done

File: contrib/babelfishpg_tsql/src/rolecmds.c (hunk 10 above)
Comment (thephantomthief) [reply contrib/babelfishpg_tsql/src/rolecmds.c:2812]: Done

File: contrib/babelfishpg_tsql/src/rolecmds.c (hunk 13 above)
Comment (thephantomthief) [reply contrib/babelfishpg_tsql/src/rolecmds.c]: Made it false

File: contrib/babelfishpg_tsql/src/rolecmds.c (hunk 14 above)
Comment (thephantomthief) [reply contrib/babelfishpg_tsql/src/rolecmds.c]: It does, but I noticed an inconsistency where sometimes we will first attempt to transfer ownership of say a procedure before its schema which was causing this operation to fail inconsistently. To workaround this, I am explicitly transferring ownership of schemas first and then running REASSIGN OWNED BY

File: contrib/babelfishpg_tsql/src/rolecmds.c (hunk 15 above)
Comment (thephantomthief) [reply contrib/babelfishpg_tsql/src/rolecmds.c:2845]: For read-only catalog access, AccessShareLock should suffice right?

File: contrib/babelfishpg_tsql/src/rolecmds.c (hunk 11 above)
Comment (thephantomthief) [reply contrib/babelfishpg_tsql/src/rolecmds.c:3215]: Done

File: contrib/babelfishpg_tsql/src/hooks.c [hunk 18]
```diff
@@ -5719,3 +5725,65 @@ handle_grantstmt_for_dbsecadmin(ObjectType objType, Oid objId, Oid ownerId,
 	}
 	return;
 }
+
+/*
+ * Objects are always owned by current user in postgres but in babelfish
+ * schema contained objects should be owned by the schema owner by default
+ * Use this hook to pick schema owner as object owner during object creation
+ * We currently only do this if current user is member of db_ddladmin
```
Comment (shalinilohia50) [inline contrib/babelfishpg_tsql/src/hooks.c]:
```
We currently only do this if current user is member of db_ddladmin
```
This should be for db_owner as well, right?

File: contrib/babelfishpg_tsql/src/rolecmds.c [hunk 19]
```diff
@@ -2605,3 +2662,458 @@ bbf_is_member_of_role_nosuper(PG_FUNCTION_ARGS)
 
 	PG_RETURN_BOOL(result);
 }
+
+char*
+get_obj_role(const char *rolname)
```
Comment (shalinilohia50) [inline contrib/babelfishpg_tsql/src/rolecmds.c:2804]: We create a role "rolname_obj" when a rolename is added as a member of db_owner. 
Are we making sure that a User cannot create this "_obj" role on its own?

File: contrib/babelfishpg_tsql/src/rolecmds.c (hunk 15 above)
Comment (shalinilohia50) [reply contrib/babelfishpg_tsql/src/rolecmds.c:2845]: Yes.

File: contrib/babelfishpg_tsql/src/rolecmds.c (hunk 16 above)
Comment (shalinilohia50) [reply contrib/babelfishpg_tsql/src/rolecmds.c:2945]: It should be `AccessShareLock` for read only.

File: contrib/babelfishpg_tsql/src/hooks.c (hunk 18 above)
Comment (thephantomthief) [reply contrib/babelfishpg_tsql/src/hooks.c]: Yeah, actually I just copy pasted this from db_ddladmin to test if the same solution will work. My changes will be rebased over db_ddladmin commit so I will fix the comment when I'll get be resolving the conflicts.

File: contrib/babelfishpg_tsql/src/rolecmds.c (hunk 19 above)
Comment (thephantomthief) [reply contrib/babelfishpg_tsql/src/rolecmds.c:2804]: We are not blocking it. There will be a limitation where if we already have say users u1_obj and u1, then end user won't be able to add u1 to db_owner role and we will throw an error. We had talked about this during design discussion and it was decided doing it like this is not really a big limitation

**Review by shalinilohia50 — APPROVED** [review]

File: contrib/babelfishpg_tsql/src/rolecmds.c [hunk 20]
```diff
@@ -2605,3 +2662,458 @@ bbf_is_member_of_role_nosuper(PG_FUNCTION_ARGS)
...
+		return;
+
+	cur_db_name = get_cur_db_name();
+	dbo_id = get_role_oid(get_dbo_role_name(cur_db_name), true);
+
+	/*
+	 * Don't change object owner if current user is
+	 * dbo or we are creating system databases
+	 */
+	if (role_oid == dbo_id || dbo_id == InvalidOid || is_create_bbf_builtin_dbs)
```
Comment (tanscorpio7) [inline contrib/babelfishpg_tsql/src/rolecmds.c]: Why is this flag needed, wouldn't we be in non tds connection when creating built in dbs ?

File: contrib/babelfishpg_tsql/src/rolecmds.c [hunk 21]
```diff
@@ -2605,3 +2662,458 @@ bbf_is_member_of_role_nosuper(PG_FUNCTION_ARGS)
...
+					   false,
+					   PROCESS_UTILITY_SUBCOMMAND,
+					   NULL,
+					   NULL,
+					   None_Receiver,
+					   NULL);
+
+		pfree(query.data);
+
+		if (obj_rolname)
```
Comment (tanscorpio7) [inline contrib/babelfishpg_tsql/src/rolecmds.c]: can this be null in some cases ?

File: contrib/babelfishpg_tsql/src/rolecmds.c [hunk 22]
```diff
@@ -2605,3 +2662,458 @@ bbf_is_member_of_role_nosuper(PG_FUNCTION_ARGS)
...
+ * db_owner role. If it is, then we will attempt drop the linked
+ * "_obj" role. If it is not, this function will do nothing.
+ */
+static void
+drop_db_owner_related_roles(Oid roleid, const char* rolname)
+{
+	Oid	db_owner_id;
+	char	*obj_rolname = NULL;
+
+	db_owner_id = get_role_oid(get_db_owner_name(get_cur_db_name()), false);
```
Comment (tanscorpio7) [inline contrib/babelfishpg_tsql/src/rolecmds.c]: can we use `get_db_owner_oid`

File: contrib/babelfishpg_tsql/src/hooks.c [hunk 23]
```diff
@@ -5807,9 +5811,11 @@ pltsql_get_object_owner(Oid namespaceId, Oid ownerId)
 			Oid 	db_owner = get_db_owner_oid(db_name, false);
 			Oid 	schema_db_id = get_dbid_from_physical_schema_name(NameStr(nsptup->nspname), false);
 
+			/* If current user is member of db_owner or db_ddladmin but not dbo */
 			if (schema_db_id == get_cur_db_id() &&
-				!has_privs_of_role(GetUserId(), db_owner) &&
+				(has_privs_of_role(GetUserId(), db_owner) ||
 				has_privs_of_role(GetUserId(), db_ddladmin))
+				&& user_oid != dbo_oid)
```
Comment (tanscorpio7) [inline contrib/babelfishpg_tsql/src/hooks.c]: This should be `ownerId` instead of user_oid. Some places will have ownerId and `GetUserId()` different even in engine and lets move this check before has_privs_of_role for efficiency,

File: contrib/babelfishpg_tsql/src/multidb.c [hunk 24]
```diff
@@ -268,11 +253,27 @@ rewrite_object_refs(Node *stmt)
 							(errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
 							 errmsg("Cannot use the special principal '%s'", principal_name)));
 
+				physical_principal_name = get_physical_user_name(db_name, principal_name, false, true);
+
+				/* Forbidden ALTER ROLE db_owner ADD/DROP MEMBER if MEMBER is a T-SQL database role */
+				if ((strcmp(role_name, "db_owner") == 0) && get_db_principal_kind(get_role_oid(physical_principal_name, false), db_name) == BBF_ROLE)
```
Comment (tanscorpio7) [inline contrib/babelfishpg_tsql/src/multidb.c]: MACRO for "db_owner" is defined.

File: contrib/babelfishpg_tsql/src/pl_handler.c (hunk 6 above)
Comment (tanscorpio7) [reply contrib/babelfishpg_tsql/src/pl_handler.c]: MACRO for `(ALTER ROLE ADD )`

File: contrib/babelfishpg_tsql/src/pl_handler.c [hunk 25]
```diff
@@ -3742,6 +3751,13 @@ bbf_ProcessUtility(PlannedStmt *pstmt,
 							create_schema->authrole = NULL;
 							alter_owner = true;
 						}
+
+						if (has_privs_of_role(owner_oid, get_db_owner_oid(db_name, false)) && owner_oid != get_role_oid(get_dbo_role_name(db_name), false))
```
Comment (tanscorpio7) [inline contrib/babelfishpg_tsql/src/pl_handler.c]: Let us have a helper for getting the dbo role oid given the db name. This will help reduce changes in multiple places.

**Review by tanscorpio7 — CHANGES_REQUESTED** [review]

File: contrib/babelfishpg_tsql/src/rolecmds.c (hunk 15 above)
Comment (HarshLunagariya) [reply contrib/babelfishpg_tsql/src/rolecmds.c:2845]: I was suggesting this because this row is being changed here -- We are changing the owner. Anyways, Engine API will anyway hold row ex. lock internally. so no worries to handle it here.

**Review by HarshLunagariya — COMMENTED** [review]
Code changes looks good. There are failures in some of workflows.

File: contrib/babelfishpg_tsql/src/rolecmds.c (hunk 20 above)
Comment (thephantomthief) [reply contrib/babelfishpg_tsql/src/rolecmds.c]: Yeah TDS_CONN check should suffice

File: contrib/babelfishpg_tsql/src/rolecmds.c (hunk 21 above)
Comment (thephantomthief) [reply contrib/babelfishpg_tsql/src/rolecmds.c]: No. Removed the if()

File: contrib/babelfishpg_tsql/src/rolecmds.c (hunk 22 above)
Comment (thephantomthief) [reply contrib/babelfishpg_tsql/src/rolecmds.c]: Yes. Done

File: contrib/babelfishpg_tsql/src/hooks.c (hunk 23 above)
Comment (thephantomthief) [reply contrib/babelfishpg_tsql/src/hooks.c]: Done

File: contrib/babelfishpg_tsql/src/multidb.c (hunk 24 above)
Comment (thephantomthief) [reply contrib/babelfishpg_tsql/src/multidb.c]: Replaced with DB_OWNER

File: contrib/babelfishpg_tsql/src/pl_handler.c (hunk 25 above)
Comment (thephantomthief) [reply contrib/babelfishpg_tsql/src/pl_handler.c]: Done

File: contrib/babelfishpg_tsql/src/hooks.c (hunk 3 above)
Comment (thephantomthief) [reply contrib/babelfishpg_tsql/src/hooks.c:2983]: Added for more object types

File: contrib/babelfishpg_tsql/src/dbcmds.c [hunk 26]
```diff
@@ -342,16 +342,18 @@ gen_dropdb_subcmds(const char *dbname, List *db_users)
 	foreach(elem, db_users)
 	{
 		char	   *user_name = (char *) lfirst(elem);
-		char 	   *original_user_name = get_authid_user_ext_original_name(user_name, dbname);
+		char 	   *original_user_name = get_authid_user_ext_original_name(user_name, dbname, true);
```
Comment (HarshLunagariya) [inline contrib/babelfishpg_tsql/src/dbcmds.c:345]: When do we get original_user_name NULL?

File: contrib/babelfishpg_tsql/src/pl_handler.c [hunk 27]
```diff
@@ -3069,7 +3069,15 @@ bbf_ProcessUtility(PlannedStmt *pstmt,
...
-							create_bbf_authid_user_ext(stmt, isuser, isuser, from_windows);
+
+							/*
+							 * If the role is created internally as part of ALTER ROLE
+							 * db_owner ADD MEMBER ... statement, we should not add this to
+							 * our babelfish catalog. These roles are meant to be internal
+							 * and not be visible to customer from Babelfish endpoint.
+							 */
+							if (strcmp(queryString, INTERNAL_ALTER_ROLE) != 0)
+								create_bbf_authid_user_ext(stmt, isuser, isuser, from_windows);
```
Comment (HarshLunagariya) [inline contrib/babelfishpg_tsql/src/pl_handler.c:3080]: Let's track the effort somewhere to add this constraint that "if any user is member of db_owner then it should have _obj user created" in metadata consistency check.

File: test/JDBC/expected/db_owner-vu-verify.out [hunk 28]
```diff
@@ -0,0 +1,2009 @@
...
+JOIN pg_roles r ON (m.roleid = r.oid)
+JOIN pg_roles mr ON (m.member = mr.oid)
+WHERE mr.rolname = 'dbowner__main_db_dbowner__u1'
+ORDER BY r.rolname;
+go
+~~START~~
+name
+dbowner__main_db_db_owner
+dbowner__main_db_dbo
+dbowner__main_db_dbowner__u1_obj
```
Comment (HarshLunagariya) [inline test/JDBC/expected/db_owner-vu-verify.out:26]: Have we tested `is_member('dbowner__main_db_dbowner__u1', 'dbo')` from ideal behaviour P.O.V. ?

Comment (HarshLunagariya) [reply test/JDBC/expected/db_owner-vu-verify.out:26]: + just double checking if we have existing case covering for sp_helpuser, sp_helprolemember,... etc procedures.
+ We need to test extended properties as well.

**Review by HarshLunagariya — APPROVED** [review]

**Review by tanscorpio7 — APPROVED** [review]
